Packet++/src/PacketUtils.cpp

```diff
@@ -17,30 +18,35 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
 	for (size_t i = 0; i<vecSize; i++)
 	{
 		uint32_t localSum = 0;
-		size_t buffLen = vec[i].len;
-		while (buffLen > 1)
+
+		// vec len is in bytes
+		for (size_t j = 0; j < vec[i].len / 2; j++)
 		{
-			PCPP_LOG_DEBUG("Value to add = 0x" << std::uppercase << std::hex << *(vec[i].buffer));
-			localSum += *(vec[i].buffer);
-			++(vec[i].buffer);
-			buffLen -= 2;
+			PCPP_LOG_DEBUG("Value to add = 0x" << std::uppercase << std::hex << vec[i].buffer[j]);
+			localSum += vec[i].buffer[j];
 		}
 		PCPP_LOG_DEBUG("Local sum = " << localSum << ", 0x" << std::uppercase << std::hex << localSum);
 
-		if (buffLen == 1)
-		{
-			uint16_t lastByte = 0;
-			*((uint8_t*)(&lastByte)) = *((uint8_t*)(vec[i].buffer));
+		// check if there is one byte left
+		if (vec[i].len % 2) {
+			// access to the last byte using an uint8_t pointer
+			uint8_t *vecBytes = (uint8_t *)vec[i].buffer;
+			uint8_t lastByte = vecBytes[vec[i].len - 1];
 			PCPP_LOG_DEBUG("1 byte left, adding value: 0x" << std::uppercase << std::hex << lastByte);
-			localSum += lastByte;
+			// We have read the latest byte manually but this byte should be interpreted
+			// as a 0xFF on LE and a 0xFF00 on BE to have a proper checksum computation
+			uint16_t lastSum = 0;
+			memcpy(&lastSum, &lastByte, 1);
+			localSum += lastSum;
+
 			PCPP_LOG_DEBUG("Local sum = " << localSum << ", 0x" << std::uppercase << std::hex << localSum);
 		}
 
+		// carry count is added to the sum
 		while (localSum>>16)
 		{
 			localSum = (localSum & 0xffff) + (localSum >> 16);
 		}

```

Comment:
btw. why do we do the carry twice? once on the local sum and once on sum. Should be enough to do it only on sum.

please have a look at the original version of tcpdump. I very much like the simplicity of it:
https://github.com/the-tcpdump-group/tcpdump/blob/a91d20cc8491f2afcc04e51474244710481244b9/print-ip.c#L276

also the implementation and comment of the next function is notable:
https://github.com/the-tcpdump-group/tcpdump/blob/a91d20cc8491f2afcc04e51474244710481244b9/print-ip.c#L312

Reply:
I think I was wrong. Not adding the carrys of local sum will give us problems for very long messages.

Reply:
Yes, we need to add the carry 😃